src/ripple/core/SQLInterface.h

```diff
@@ -0,0 +1,259 @@
+#include <boost/variant.hpp>
+
+namespace ripple {
+
+class SQLInterface
+{
+public:
+    enum DatabaseType : uint8_t { LEDGER, TRANSACTION };
+
+    enum TableType { LEDGERS, TRANSACTIONS, ACCOUNT_TRANSACTIONS };
+
+    struct SQLLedgerInfo
+    {
+        boost::optional<std::string> sLedgerHash, sPrevHash, sAccountHash,
+            sTransHash;
+        boost::optional<std::uint64_t> totDrops, closingTime, prevClosingTime,
+            closeResolution, closeFlags, ledgerSeq64;
+    };
+
+    typedef std::pair<LedgerIndex, uint8_t> DatabaseIndex;
+
+    struct hash_pair
+    {
+        template <class T1, class T2>
+        size_t
+        operator()(const std::pair<T1, T2>& p) const
+        {
+            auto hash1 = std::hash<T1>{}(p.first);
+            auto hash2 = std::hash<T2>{}(p.second);
+            return hash1 ^ hash2;
+        }
+    };
+
+    static std::string
+    tableName(TableType type)
+    {
+        switch (type)
+        {
+            case LEDGERS:
+                return "Ledgers";
+            case TRANSACTIONS:
+                return "Transactions";
+            case ACCOUNT_TRANSACTIONS:
+                return "AccountTransactions";
+        }
+        return "Unknown";
+    }
+
+    static std::unique_ptr<SQLInterface>
+    init(Application& app, Config const& config);
+
+    virtual bool
+    makeLedgerDBs(
+        Config const& config,
+        beast::Journal const& j,
+        bool setupFromConfig,
+        LedgerIndex shardIndex,
+        bool backendComplete,
+        boost::filesystem::path const& dir,
+        DatabaseCon::CheckpointerSetup const& checkpointerSetup) = 0;
+
+    virtual void
+    addDBs(LedgerIndex shardIndex, DatabaseCon* tx, DatabaseCon* lgr) = 0;
+
+    virtual void
+    removeDBs(LedgerIndex shardIndex) = 0;
+
+    virtual ~SQLInterface()
+    {
+    }
+
+    virtual boost::optional<LedgerIndex>
+    getMinLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual boost::optional<LedgerIndex>
+    getMaxLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual void
+    deleteByLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual void
+    deleteBeforeLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual int
+    getRows(DatabaseType dbtype, TableType type) = 0;
+
+    virtual std::tuple<int, int, int>
+    getRowsMinMax(DatabaseType dbtype, TableType type) = 0;
+
+    virtual bool
+    saveValidatedLedger(
+        Application& app,
+        std::shared_ptr<Ledger const> const& ledger,
+        bool current) = 0;
+
+    virtual bool
+    loadLedgerInfoByIndex(
+        SQLInterface::SQLLedgerInfo& info,
+        beast::Journal const& j,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual bool
+    loadLedgerInfoByIndexSorted(
+        SQLInterface::SQLLedgerInfo& info,
+        beast::Journal const& j,
+        bool ascendSort) = 0;
```

Comment:
`ascendSort` is an implementation detail that doesn't belong on the API. It really reamns "return the ledger with the smallest (or largest) ledger seq". A couple of questions:
1) Do we actually need this parameter at all? (I only found uses for the newest, tho I may have missed the calls for the oldest). 
2) If we do need this, could the public interface be replaced with two functions: "loadOldestLedgerInfo" or "loadNewestLedgerInfo", which then call an internal helper function that takes the bool param?

Ditto for the other "bySorted" functions

Reply:
I vote for removing the bool param and having two separate functions

Reply:
Done.

Reply:
Do we need `loadOldestLedgerInfo`? I don't see any callers, did I miss them? If that function isn't used I'd remove it (unless we know we're going to need it in the near future).

Reply:
Done. Removed.

Reply:
`loadOldestLedgerInfo` still needs to be removed from `SQLInterface.md`, `SQLInterface_nodes.h` and `SQLInterface_nodes.cpp`

Reply:
Done.